docs/rfcs/0020-rotate-hash-peppers.md

```diff
@@ -0,0 +1,384 @@
+
+If BFD wishes to maintain total control over hashing of URL GET parameters, we could adopt asymmetric crypto for creating the hash value(s). Instead of _sharing_ a secret (the algorithm), we use a public-private keypair to dynamically encrypt/decrypt sensitive URL parameters.
+
+BFD creates a public-private keypair; it then distributes the public key to each of its peering partners. Each PP then uses their copy of the public key to encrypt a plaintext MBI or HICN; the encrypted value is then passed to the BFD _patient search_ as a URL GET parameter. The parameter is fully obfuscated so any public access (logging, TCP snooping, etc.) is limited to just that...a URL that has an encrypted content and the only way to discern the view the sensitive parameter is to decrypt it using BFD's private key.
+##### Benefits
+- can still support HTTP GET URLs; current logging continues as is, logging the URL.
+- extensible choice of algorithms and key sizes make this the safest way to protect data.
+- wide range of algorithms and keysizes make this extremely extensible and bullet-proof.
+- already using asymmetric in our TLS implementation.
+- completely removes necessity to have/maintain hash algorithms and hashed data values; completely removes necessity to store hashed data (_MBI_HASH_ and _BENE_CRNT_HIC_NUM_ values); lookups now occur vs. actual _MBI_NUM_ or HICN_UNHASHED_ value.
+##### Drawbacks
+- slight overkill for what BFD needs.
+- feeling performance will take a hit, but this can be mitigated by using Elliptical Curve Cryptograph (ECC) which provides better performance by using smaller keysizes which are just as resilient as RSA that depends on larger keysizes.
+- small learning curve for some devs, but BFD implementation would not be complex.
+
+
+## Prior Art
+[Prior Art]: #prior-art
+While BFD has no prior work replacing HTTP GET with HTTP POST, the FHIR specification is clear on how this could be implemented.
+
+RDA (PACA) implemented a hashed MBI cache table to their database schema which handles current and previous (old) MBI hash value(s) within their patient lookup queries. If a client requesting pre-adjudicated claims data passes in an old hashed value, and that value exists in the cache table, the query lookup succeeds **AND** the patient's current MBI hash value is added to the cache table if it does not already exist there. In the case of a hash algorithm change, no extra work would be necessary to maintain the cache, as it dynamically maintains itself. Should the adoption of HTTP POST be deferred, this cache table strategy could be adopted for adjudicated claims.
```

Comment:
The table is not updated when a client makes a call.  It would have to be updated in a separate process.

On the API side the mbi cache table supports looking up an MBI from either its old or current MBI hash value.  This is to allow the server to continue serving requests with the old hash values while a background process is updating the new hash values in the table.

The remediation process would be:

* Stop the pipeline
* `UPDATE rda.mbi_cache SET old_hash=hash`
* Restart the `bfd-server` with `-DPacOldMbiHashEnabled=true`
* In a background process scan the `mbi_cache` table and set the `hash` column to the new hash value.
* Once all updates have been completed restart the pipeline.
* Some time later when all clients have begun using the new hash we can restart the server with `-DPacOldMbiHashEnabled=false` and `UPDATE mbi_cache SET old_hash=null`.

Reply:
modified to reflect your (much better!) description; thx!